tests/integration_test.go

```diff
@@ -0,0 +1,272 @@
+		mux.ServeHTTP(w, r)
+	})
+
+	httpSvr := httptest.NewServer(f)
+	t.Cleanup(httpSvr.Close)
+	return Addrs{
+		GRPCEndpoint: ln.Addr().String(),
+		HTTPEndpoint: httpSvr.Listener.Addr().String(),
+	}
+}
+
+func Test_OTLPHTTP(t *testing.T) {
+	totalMetric := atomic.Int64{}
+	addrs := NewTestServer(t, func(req *otlpcolmetrics.ExportMetricsServiceRequest) {
+		for _, resourceMetric := range req.GetResourceMetrics() {
+			for _, scopeMetric := range resourceMetric.GetScopeMetrics() {
+				count := int64(len(scopeMetric.GetMetrics()))
+				totalMetric.Add(count)
+			}
+		}
+	})
+	envoyPort := portal.New(t).One()
+	hclog.Default().Info("Running test server", "addr", addrs)
+	collector, err := otel.NewCollector(otel.CollectorCfg{
+		ExporterConfig: &config.ExporterConfig{
+			ID: exporters.BaseOtlpExporterID,
+			Exporter: &exporters.ExporterConfig{
+				Endpoint: fmt.Sprintf("http://%s", addrs.HTTPEndpoint),
+				Headers: map[string]string{
+					"authorization": "abc123",
+				},
+			},
+		},
+		MetricsPort:  portal.New(t).One(),
+		BatchTimeout: time.Second,
+		EnvoyPort:    envoyPort,
+	})
+	must.NoError(t, err)
+	ctx := context.Background()
+	go func() { must.NoError(t, collector.Run(ctx)) }()
+
+	total := generateMetrics(t, envoyPort, 30, 30)
+	ch := time.After(30 * time.Second)
+	for {
+		select {
+		case <-ch:
+			t.Logf("Failed to get %d metrics in 30 seconds", total)
+			t.Fail()
+		default:
+			t.Log(totalMetric.Load(), total)
+			time.Sleep(500 * time.Millisecond)
+		}
+		if totalMetric.Load() == int64(total) {
+			break
+		}
+	}
+
+	collector.Shutdown()
+	hclog.Default().Info("Shutting down")
+}
+
+func Test_OTLPGRPC(t *testing.T) {
+	totalMetric := atomic.Int64{}
+	addrs := NewTestServer(t, func(req *otlpcolmetrics.ExportMetricsServiceRequest) {
+		for _, resourceMetric := range req.GetResourceMetrics() {
+			for _, scopeMetric := range resourceMetric.GetScopeMetrics() {
+				count := int64(len(scopeMetric.GetMetrics()))
+				totalMetric.Add(count)
+			}
+		}
+	})
+
+	envoyPort := portal.New(t).One()
+	hclog.Default().Info("Running test server", "addr", addrs)
+	collector, err := otel.NewCollector(otel.CollectorCfg{
+		ExporterConfig: &config.ExporterConfig{
+			ID: exporters.GRPCOtlpExporterID,
+			Exporter: &exporters.ExporterConfig{
+				Endpoint: fmt.Sprintf("http://%s", addrs.GRPCEndpoint),
+				Headers: map[string]string{
+					"authorization": "abc123",
+				},
+			},
+		},
+		MetricsPort:  portal.New(t).One(),
+		BatchTimeout: time.Second,
+		EnvoyPort:    envoyPort,
+	})
+	must.NoError(t, err)
+	ctx := context.Background()
+	go func() { must.NoError(t, collector.Run(ctx)) }()
+
+	total := generateMetrics(t, envoyPort, 30, 30)
+	ch := time.After(30 * time.Second)
+	for {
+		select {
+		case <-ch:
+			t.Logf("Failed to get %d metrics in 30 seconds", total)
+			t.Fail()
+		default:
+			t.Log(totalMetric.Load())
+			time.Sleep(500 * time.Millisecond)
+		}
+		if totalMetric.Load() == int64(total) {
```

Comment:
Just a note that we had a lengthy discussion about this test, I think we should add a TODO here for now about the actual validation aspect. Maybe this is done in the future, regardless this test is a step forward from our discussions. So I won't block on this!